File: booking/www/reservation.py
# -*- coding: utf-8 -*-
# Copyright (c) 2015, Britlog and contributors
# For license information, please see license.txt

from __future__ import unicode_literals

import frappe
import datetime

no_cache = 1

def get_context(context):
	# do your magic here
	pass

@frappe.whitelist(allow_guest=True)
def get_stock(slot):
	stock = frappe.get_doc("Booking Slot", slot)

	return str(stock.available_places)

@frappe.whitelist(allow_guest=True)
def get_introduction():
	return frappe.db.get_single_value('Booking Settings', 'booking_form_introduction')

@frappe.whitelist(allow_guest=True)
def get_activities():

	return frappe.db.sql("""
		select	distinct BS.type 
		from `tabBooking Slot` BS
		where time_slot > NOW() and show_in_website = 1
		order by BS.type""")

@frappe.whitelist(allow_guest=True)
def get_slots(activity):

	slots = frappe.db.sql("""
		select	name, time_slot_display, type, location, available_places, total_places,
			practical_information, is_replay 
		from `tabBooking Slot`
		where (time_slot > NOW() or is_replay = 1) and show_in_website = 1
			  and type = case when %(activity)s != '' then %(activity)s else type end
		order by time_slot asc""",{"activity": activity}, as_dict=True)

	for slot in slots:
		slot['subscription_places'] = slot.get('total_places') \
			- frappe.db.sql("""select COUNT(*) from `tabBooking Subscriber` where parent = %(slot)s""",
			{"slot": slot.get('name')})[0][0]

	return slots

@frappe.whitelist(allow_guest=True)
def set_notification(slot, email, name):
	doc = frappe.get_doc("Booking Slot", slot)

	# check if already registered
	booked = frappe.db.sql("""select COUNT(*)
	                from `tabBooking`
	                where `tabBooking`.slot = %(slot)s and `tabBooking`.email_id = %(email)s""",
						   {"slot": slot, "email": email})[0][0]

	booked += frappe.db.sql("""select COUNT(*)
	                from `tabBooking Subscriber`
	                inner join `tabCustomer` on `tabBooking Subscriber`.subscriber = tabCustomer.name
	                inner join `tabDynamic Link` on tabCustomer.name=`tabDynamic Link`.link_name
	                inner join `tabContact` on `tabDynamic Link`.parent=tabContact.name
	                where `tabBooking Subscriber`.parent = %(slot)s and `tabContact`.email_id = %(email)s""",
							{"slot": slot, "email": email})[0][0]

	# raise error
	if booked > 0:
		frappe.throw("Vous êtes déjà inscrit à cette séance")
	else:

		# check if already on waiting list
		waiting_list = frappe.db.sql("""select COUNT(*)
			                from `tabBooking Notification`
			                where parent = %(slot)s and email_id = %(email)s""",
					  {"slot": slot, "email": email})[0][0]

		if waiting_list > 0:
			frappe.throw("Vous êtes déjà inscrit sur la liste d'attente de cette séance")
		else:
			doc.append("notifications", {
				"email_id": email,
				"full_name": name,
				"request_date": datetime.datetime.now()
			})
			doc.save()

@frappe.whitelist(allow_guest=True)
def add_booking(slot, email, name, city, phone, sms, comment):
	validation = {}

	doc = frappe.get_doc({
		"doctype": "Booking",
		"slot": slot,
		"email_id": email,
		"full_name": name,
		"city": city,
		"phone": phone,
		"confirm_sms": sms == 'true',
		"comment": comment
	})
	doc.insert()

	# payment URL
	validation["payment_url"] = frappe.get_value("Payment Request", doc.payment_request, 'payment_url')
	validation["success_message"] = "Réservation confirmée"

	return validation

File: booking/www/reservation.js

frappe.ready(function() {
	// bind events here

	// load introduction text
	frappe.call({
        method: 'booking.www.reservation.get_introduction',
        args: {},
        callback: function(r) {
            if (r.message){
                $('.introduction p').html(r.message);
            };
        }
    });

	// load activities
	$('[name="type"]').empty();
	$('[name="type"]').append($('<option>').val('').text('- Toutes -'));
	frappe.call({
			method: 'booking.www.reservation.get_activities',
			args: {},
			callback: function(r) {
//				console.log(r.message);

				(r.message || []).forEach(function(row){
					$('[name="type"]').append($('<option>').val(row).text(row));
				});

			}
		});

	// load slots
    get_slots("");

	function get_slots(activity) {

		$('[name="slot"]').empty();
    	$('[name="slot"]').append($('<option>').val('').text(''));

		frappe.call({
			method: 'booking.www.reservation.get_slots',
			args: {
				'activity': activity
			},
			callback: function(r) {
//				console.log(r.message);
				var available_message = "";
				(r.message || []).forEach(function(row){
					if (row.available_places <= 0) {
						available_message = "Liste d'attente";
					}
					else if (row.available_places == 1) {
						available_message = "1 place disponible";
					}
					else {
						available_message = row.available_places+" places disponibles";
					}

					$('[name="slot"]').append($('<option>').val(row.name).text((row.time_slot_display || row.name)
						+((activity) ? '' : " | "+row.type.toUpperCase())
						+((row.location) ? ' '+row.location : '')
						+((row.is_replay) ? ' REPLAY' : '')
						+" | "+available_message)
						.attr('available_places',row.available_places).attr('subscription_places',row.subscription_places)
						.attr('practical_information',row.practical_information));

	//                if (row.available_places == 0) { $('select option:contains("'+row.name+'")').attr("disabled", "disabled"); }
				});

			}
		});
    }

    function valid_email_fr(email) {
	    var re = /^(([^<>()[\]\\.,;:\s@\"]+(\.[^<>()[\]\\.,;:\s@\"]+)*)|(\".+\"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
  		return re.test(email);
    }

    function valid_phone(id,mobile) {
    if (mobile)
        return (id.search("^0[6-7][0-9]{8}$")==-1) ? 0 : 1;
    else
        return (id.search("^0[1-9][0-9]{8}$")==-1) ? 0 : 1;
    }

	function display_payment() {
		// Check subscription validity if any
		frappe.call({
            method: 'booking.booking.doctype.booking.booking.get_slot_subscription',
            args: {
                'email_id': $('[name="email_id"]').val(),
                'slot_id': $('[name="slot"]').val()
            },
            callback: function(r) {
//                console.log(r.message);

				if (jQuery.isEmptyObject(r.message) || !r.message.is_valid) {

					if (r.message.price) {
						$('#amount').html("Montant à payer : "+r.message.price);
					} else {
						$('#amount').html("");
					}

					if (r.message.to_bill) {
						$("#credit-card").prop("checked", true);
						$('#credit-card-group').show();
					} else {
						$('#credit-card-group').hide();
					}

					if (r.message.allow_payment_on_site) {
						if (r.message.payment_instruction) {
							$('label[for=payment-on-site]').html(r.message.payment_instruction);
						} else {
							$('label[for=payment-on-site]').html("Paiement sur place");
						}
						if (!r.message.to_bill) {
							$("#payment-on-site").prop("checked", true);
						}
						$('#payment-on-site-group').show();
					} else {
						$('#payment-on-site-group').hide();
					}

					$('#payment-group').show();

				} else $('#payment-group').hide();
            }
        });
    }

    $('[name="slot"]').change(function () {

		  // display practical information if specified
		  if ($('[name="slot"] :selected').attr('practical_information')) {
		  	$("#practical-information").html($('[name="slot"] :selected').attr('practical_information'));
		  	$("#practical-information").toggle(true);
          }
          else {
            $("#practical-information").html('');
            $("#practical-information").toggle(false);
          }

		  // manage notification button display
          $("#notification-button").prop("disabled",false);

          if ($('[name="slot"] :selected').attr('available_places') <= 0) {
            $("#notification-button").toggle(true);		//show button
          }
          else {
            $("#notification-button").toggle(false);	//hide button
          }

		  if ($('[name="slot"] :selected').attr('subscription_places') <= 0) {
		  	frappe.msgprint("Le groupe est complet, cours à la séance uniquement.");
		  }

		  // email already entered and slot is changed => display payment based on the selected activity
		  if ($('[name="email_confirm"]').val()) {
		  	display_payment();
		  }
     })

	$('[name="type"]').change(function () {

//		  frappe.msgprint($('[name="type"]').val());
//		  $('[name="slot"]').val($('[name="slot"] option:first').val());
		  $("#notification-button").toggle(false);	//hide button

		  // reload slots
		  get_slots($('[name="type"]').val());

		  $("#practical-information").html('');
          $("#practical-information").toggle(false);

     })

    $('#notification-button').on("click", function() {
        var email = $('[name="email_id"]').val();
        var email_confirm = $('[name="email_confirm"]').val();
        var fullname = $('[name="full_name"]').val();
        var slot = $('[name="slot"]').val();

        if (!email || !valid_email_fr(email)) {
            frappe.msgprint(__("Entrez s'il vous plaît une adresse e-mail valide, sans accents ni espaces."));
            $('[name="email_id"]').focus();
            return false;
		}

		if(email != email_confirm) {
			frappe.msgprint(__("Veuillez contrôler votre adresse e-mail : les champs E-mail et Confirmation E-mail sont différents."));
            $('[name="email_confirm"]').focus();
            return false;
		}

		if (!fullname) {
            frappe.msgprint(__("Entrez s'il vous plaît votre nom."));
            $('[name="full_name"]').focus();
            return false;
		}

		frappe.call({
            method: 'booking.www.reservation.set_notification',
            args: {
                'slot': slot,
                'email': email,
                'name': fullname
            },
            callback: function(r) {
                //console.log(r.message);
                frappe.msgprint(__("Votre demande d'inscription sur la liste d'attente est bien enregistrée.\
                	Vous allez recevoir un e-mail si une place se libère."));
                $("#notification-button").prop("disabled",true);
            }
        });
    });

	$('[name="email_confirm"]').change(function () {
		  display_payment();
     })

    $('.btn-form-submit').on("click", function() {
		var slot = $('[name="slot"]').val();
		var fullname = $('[name="full_name"]').val();
		var email = $('[name="email_id"]').val();
		var email_confirm = $('[name="email_confirm"]').val();
		var city = $('[name="city"]').val();
		var phone = $('[name="phone"]').val();
		var sms = $('[name="confirm_sms"]').is(':checked');
		var comment = $('[name="comment"]').val();
		var payment = $('[name="payment"]:checked').val();

		if(!slot) {
			frappe.msgprint(__("Veuillez choisir un cours."));
            $('[name="slot"]').focus();
            return false;
		}

		if(!fullname) {
			frappe.msgprint(__("Veuillez saisir votre nom."));
            $('[name="full_name"]').focus();
            return false;
		}

		if(!valid_email_fr(email)) {
            frappe.msgprint(__("Entrez s'il vous plaît une adresse e-mail valide, sans accents ni espaces."));
            $('[name="email_id"]').focus();
            return false;
		}

		if(email != email_confirm) {
			frappe.msgprint(__("Veuillez contrôler votre adresse e-mail : les champs E-mail et Confirmation E-mail sont différents."));
            $('[name="email_confirm"]').focus();
            return false;
		}

		if(phone || sms) {
            phone = phone.replace(/\D/g,"");    // delete all non-numeric characters
            $('[name="phone"]').val(phone);
            if (!valid_phone(phone,sms)) {
                frappe.msgprint(__("Entrez s'il vous plaît un numéro de téléphone valide.\
                    Un numéro de portable est nécessaire si vous avez demandé une confirmation par SMS."));
                $('[name="phone"]').focus();
                return false;
            }
		}

		// Check subscription validity if any
		frappe.call({
            method: 'booking.booking.doctype.booking.booking.get_slot_subscription',
            args: {
                'email_id': email,
                'slot_id': $('[name="slot"]').val()
            },
            callback: function(r) {
                console.log(r.message);
                var bCancel = false

				if (!jQuery.isEmptyObject(r.message) && typeof r.message.is_valid !== 'undefined' && !r.message.is_valid) {
					// subscription is not valid, display a warning
					if (!confirm(r.message.warning_msg)) {
						bCancel = true;
					}
				}

                if (!bCancel) {
                	//Add the booking
					frappe.call({
						method: 'booking.www.reservation.add_booking',
						args: {
							'slot': slot,
							'email': email,
							'name': fullname,
							'city': city,
							'phone': phone,
							'sms': sms,
							'comment': comment,
							'payment_mode': payment
						},
						callback: function(r) {
							if (!jQuery.isEmptyObject(r.message)){

								if (r.message.payment_url) {
									// redirect to payment
									window.location.href = r.message.payment_url;
								}
								else {
									frappe.msgprint(r.message.success_message);

									//Init form
									$('[name="type"]').val('');
									get_slots("");
									$('[name="comment"]').val('');
								}
							};
						}
    				});
                }
            }
        });

    	return false;
	});

})


